predicators/approaches/grammar_search_invention_approach.py

```diff
@@ -442,6 +501,66 @@ def _get_feature_ranges(
         return feature_ranges
 
 
+@dataclass(frozen=True, eq=False, repr=False)
+class _FeatureDiffInequalitiesPredicateGrammar(
+        _SingleFeatureInequalitiesPredicateGrammar):
+    """Generates features of the form "|0.feature - 1.feature| >= c" or
+    "|0.feature - 1.feature| <= c". This is also capable of
+    generating single-feature inequalities if 0 and 1 are the
+    same type."""
+
+    def enumerate(self) -> Iterator[Tuple[Predicate, float]]:
+        # Get ranges of feature values from data.
+        feature_ranges = self._get_feature_ranges()
+        # Edge case: if there are no features at all, return immediately.
+        if not any(r for r in feature_ranges.values()):
+            return
+        # 0.5, 0.25, 0.75, 0.125, 0.375, ...
+        constant_generator = _halving_constant_generator(0.0, 1.0)
+        for constant_idx, (constant, cost) in enumerate(constant_generator):
+            for (t1, t2) in itertools.combinations_with_replacement(
+                    sorted(self.types), 2):
+                for f1 in t1.feature_names:
+                    for f2 in t2.feature_names:
+                        # If the types and features are the same, then
```

Comment:
why? isn't it possible that we would want to compare the same feature in two different objects of the same type?